### PR TITLE
23 make a checklist for developers

### PR DESCRIPTION
**Pull request by KoomenErina**
added documentation relevant for developers and reviewers. 
Also added general standards, conventioins, and tools.

### CODE REVIEW

File: docs/source/checklist_for_review.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+- [ ] Understand the scope of the pull request. 
+- [ ] Understand the issues related to this pull request. 
+- [ ] Clarify the scope of the review.
+
+The following sub list should be checked in order. If one sub list is not complete, request changes before reviewing any further. 
+After changes have been submitted, start with the first sublist again. 
+
+## 1. Global scan
+- [ ] Only relevant files have been included in the pull request
+- [ ] You know what files you need to review.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: What's the difference between 1 and 2?

File: docs/source/checklist_for_review.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+- [ ] The pull request only includes the requested feature. 
+- [ ] All merge conflicts have been resolved.
+- [ ] If you pull the feature branch and run all the test, they all pass without warnings. 
+- [ ] If you pretend to be an user and use this feature, it works as expected.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: There should be scenario(s) for user testing in the issue. The reviewer can propose more, but she should at least run pre-defined tests.

File: docs/source/checklist_for_review.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+- [ ] All merge conflicts have been resolved.
+- [ ] If you pull the feature branch and run all the test, they all pass without warnings. 
+- [ ] If you pretend to be an user and use this feature, it works as expected. 
+- [ ] The documentations builds without errors. 
+
+
+## 2. Reviewing the tests
+
+- [ ] Every aspect of the new feature is being taken up in the unit tests. Standard aspects include:
+    - correct shape and datatype of return value (correct dataframe/series, number of columns, datatype of columns, number of rows, column names)
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: Why? We want to use typehints by default right?

File: docs/source/checklist_for_review.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    - behaviour in "trivial" cases (input is empty dataframe, zero, None, etc...)
+    - behaviour in realistic use case. 
+    - edge cases. (values equal to border values, None, etc...)
+- [ ] unit tests are still lightweight and fully automatic.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: What does lightweight mean? Maybe give a soft maximum number of rows?

File: docs/source/checklist_for_review.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+- [ ] Every aspect of the new feature is being taken up in the unit tests. Standard aspects include:
+    - correct shape and datatype of return value (correct dataframe/series, number of columns, datatype of columns, number of rows, column names)
+    - behaviour in "trivial" cases (input is empty dataframe, zero, None, etc...)
+    - behaviour in realistic use case. 
+    - edge cases. (values equal to border values, None, etc...)
+- [ ] unit tests are still lightweight and fully automatic. 
+For each unit test:
+- [ ] The test is easy to understand.
+- [ ] The test leaves the computer the way it found it. The test should leave no proof that it has run. This implies that a cleanup should happen, even if the test fails.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: how do we know that?

File: docs/source/checklist_for_review.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+- [ ] Every aspect of the new feature is being taken up in the unit tests. Standard aspects include:
+    - correct shape and datatype of return value (correct dataframe/series, number of columns, datatype of columns, number of rows, column names)
+    - behaviour in "trivial" cases (input is empty dataframe, zero, None, etc...)
+    - behaviour in realistic use case. 
+    - edge cases. (values equal to border values, None, etc...)
+- [ ] unit tests are still lightweight and fully automatic. 
+For each unit test:
+- [ ] The test is easy to understand.
+- [ ] The test leaves the computer the way it found it. The test should leave no proof that it has run. This implies that a cleanup should happen, even if the test fails. 
+- [ ] It is clear what the test is supposed to test.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: Same as line 33?

File: docs/source/checklist_for_review.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+- [ ] The only reason the test could fail is an error in the code being tested (and not in dependencies)
+- [ ] The tests effectively proves the claim.
+- [ ] The test is not testing too much.
```
Comment (bbonnet14) [inline docs/source/checklist_for_review.md]: Same as line 31?

File: docs/source/developer_checklist.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+## Before you begin
+- [ ] Fully understand the feature request. 
+- [ ] Fully understand the proof of concept code. 
+- [ ] Fully understand how this feature should be implemented
+- [ ] Think of any edge cases or exceptions.
+
+## During development
+
+- [ ] Work on a branch based on the develop branch.
+- [ ] Only implement the feature on this branch. Found other things that need to be done first? Make a different branch.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: "Open a new issue and make a different branch"

File: docs/source/developer_checklist.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] Apply Test Driven Development. 
+- [ ] Use the GIVEN, WHEN, THEN format for unit tests. Specify the GIVEN, WHEN, THEN in the docstring of the unit tests.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: What's that?

File: docs/source/developer_checklist.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] Apply the PEP8 standard when useful. 
+- [ ] Use British english.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: typo: British English

File: docs/source/developer_checklist.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+## Before making a pull request
+
+- [ ] You can demonstrate that it works on your machine.
+- [ ] It works on your machine like it would on a users machine.
+- [ ] All tests ar passing.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: typo: are

File: docs/source/developer_checklist.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] Warnings are resolved. 
+- [ ] The in-code docstrings accurately describe the methods and classes and how to use it.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: how to use **them**.

File: docs/source/developer_checklist.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] Any other relevant documentation has been updated.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md:28]: Can you give a (exhaustive) list of the possible documentation?

File: docs/source/developer_checklist.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] It works on your machine like it would on a users machine.
+- [ ] All tests ar passing.
+- [ ] Warnings are resolved. 
+- [ ] The in-code docstrings accurately describe the methods and classes and how to use it.
+- [ ] Any other relevant documentation has been updated. 
+- [ ] Any irrelevant files are excluded from the branch (untrack/deleted).
+
+## After making a pull request, before asking review
+- [ ] The pull requests merges into develop.
+- [ ] The title of the pull requests is descriptive and reflects issue numbers.
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: typo: request

File: docs/source/developer_checklist.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+- [ ] The description of the pull requests gives a short summary what has been changed and highlights areas that need more attention (if applicable).
```
Comment (bbonnet14) [inline docs/source/developer_checklist.md]: typo: request

Comment (bbonnet14) [inline docs/source/developer_checklist.md]: With or in addition to short summary: description of the scope, intended user
Or just a link to the right issues

**Review by bbonnet14 — COMMENTED** [review]
Nice, that's very complete! 

I mostly left open questions, you can choose if more info needs to be added in the checklists. One important point I have is that each new development should come from at least one issue, so do we have the issue template available here?

File: docs/source/developer_checklist.md (hunk 13 above)
Comment (KoomenErina) [reply docs/source/developer_checklist.md:28]: no, not yet. For now, we only have functional documentation and docstrings. In the (near?) future, we might have more elaborate examples and user guides and tutorials.

**Review by bbonnet14 — APPROVED** [review]
Looks good!
